### PR TITLE
feat(webhooks): [nan-1142] update options on environment settings to allow more use settings

### PR DESCRIPTION
**Pull request by khaliqgant**
## Describe your changes
- Updates the UI to use the external webhooks table which was already successfully migrated
- Updates types in the node-client, specifically the webhook success and error objects. 
- Updates routes to use the url and settings controller files for better validation.

Note that the environment settings for webhook columns will be removed in a future PR.

## Issue ticket number and link
NAN-1142

## Checklist before requesting a review (skip if just adding/editing APIs & templates)
- [x] I added tests, otherwise the reason is: 
- [ ] I added observability, otherwise the reason is:
- [ ] I added analytics, otherwise the reason is:

### CODE REVIEW

**Comment by TBonnin** [comment]
I'll review on Friday. Sorry about the delay

File: packages/node-client/lib/types.ts [hunk 1]
```diff
@@ -1,4 +1,14 @@
 import type { ParamsSerializerOptions } from 'axios';
+import type { NangoSyncWebhookBodySuccess, NangoSyncWebhookBodyError, NangoAuthWebhookBodySuccess, NangoAuthWebhookBodyError } from '@nangohq/types';
```
Comment (bodinsamuel) [inline packages/node-client/lib/types.ts:2]: :feels-good:

File: packages/node-client/lib/types.ts [hunk 2]
```diff
@@ -1,4 +1,14 @@
...
+
+export type {
+    NangoSyncWebhookBodySuccess,
+    NangoSyncWebhookBodySuccess as NangoSyncWebhookBody,
```
Comment (bodinsamuel) [inline packages/node-client/lib/types.ts:7]: You can put (if you want) /** @deprecated use NangoSyncWebhookBodySuccess */

File: packages/shared/lib/services/external-webhook.service.ts [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+export async function update(id: number, data: WebhookSettings): Promise<void> {
+    await db.knex.from<ExternalWebhook>('_nango_external_webhooks').where({ environment_id: id }).update({
+        on_sync_completion_always: data.alwaysSendWebhook,
+        on_auth_creation: data.sendAuthWebhook,
+        on_auth_refresh_error: data.sendRefreshFailedWebhook,
+        on_sync_error: data.sendSyncFailedWebhook
+    });
+}
+
+export async function updatePrimaryUrl(id: number, primaryUrl: string): Promise<void> {
```
Comment (bodinsamuel) [inline packages/shared/lib/services/external-webhook.service.ts]: `id` is ambiguous here, its environmentId not webhookId (comment for all methods here)

File: packages/server/lib/controllers/v1/environment/webhook/settings.ts [hunk 4]
```diff
@@ -0,0 +1,39 @@
+import { z } from 'zod';
```
Comment (bodinsamuel) [inline packages/server/lib/controllers/v1/environment/webhook/settings.ts]: updateSettings.ts or patchSettings.ts would clarify (the goal is to have one endpoint per file)

File: packages/server/lib/routes.ts [hunk 5]
```diff
@@ -168,17 +170,16 @@ web.route('/api/v1/account/admin/switch').post(webAuth, accountController.switch
 
 web.route('/api/v1/environment').get(webAuth, environmentController.getEnvironment.bind(environmentController));
 web.route('/api/v1/environment/callback').post(webAuth, environmentController.updateCallback.bind(environmentController));
-web.route('/api/v1/environment/webhook').post(webAuth, environmentController.updateWebhookURL.bind(environmentController));
-web.route('/api/v1/environment/webhook-secondary').post(webAuth, environmentController.updateSecondaryWebhookURL.bind(environmentController));
+web.route('/api/v1/environment/webhook/url/primary').patch(webAuth, updatePrimaryUrl);
+web.route('/api/v1/environment/webhook/url/secondary').patch(webAuth, updateSecondaryUrl);
```
Comment (bodinsamuel) [inline packages/server/lib/routes.ts]: imo: still seems unnecessary vs `/webhook/url/:name`

File: packages/types/lib/environment/api/webhook.ts [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+    Querystring: {
+        env: string;
+    };
+    Path: '/api/v1/environment/webhook/url/primary';
+    Body: {
+        url: string;
+    };
+    Success: {
+        url: string;
+    };
```
Comment (bodinsamuel) [inline packages/types/lib/environment/api/webhook.ts:33]: Regarding api design I would be very careful of just outputing data in the main object, if any fields is named `error` then typescript will be confused. Also it will return inconsistent data type (if you data is an array then suddenly the object is now an array).And it's easier and more consistent to always answer `{ error: Error } | { data: any }` (or something like that)

File: packages/types/lib/webhooks/api.ts [hunk 7]
```diff
@@ -0,0 +1,59 @@
+import type { AuthOperationType, AuthModeType } from '../auth/api.js';
```
Comment (bodinsamuel) [inline packages/types/lib/webhooks/api.ts:1]: I guess everything can be called API but it's confusing with other api.ts (maybe we should rename other files to http.api.ts

File: packages/server/lib/controllers/v1/environment/webhook/url.ts [hunk 8]
```diff
@@ -0,0 +1,61 @@
+import { z } from 'zod';
+import { asyncWrapper } from '../../../../utils/asyncWrapper.js';
+import type { UpdatePrimaryUrl, UpdateSecondaryUrl } from '@nangohq/types';
+import { requireEmptyQuery, zodErrorToHTTP } from '@nangohq/utils';
+import { externalWebhookService } from '@nangohq/shared';
+
+const validation = z
+    .object({
+        url: z.string()
```
Comment (bodinsamuel) [inline packages/server/lib/controllers/v1/environment/webhook/url.ts]: z.string().url()

File: packages/shared/lib/services/external-webhook.service.ts [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+import type { ExternalWebhook, WebhookSettings } from '@nangohq/types';
+
+export async function get(id: number): Promise<ExternalWebhook | null> {
+    const result = await db.knex.select('*').from<ExternalWebhook>('_nango_external_webhooks').where({ environment_id: id }).first();
+
+    return result || null;
+}
+
+export async function update(id: number, data: WebhookSettings): Promise<void> {
+    await db.knex.from<ExternalWebhook>('_nango_external_webhooks').where({ environment_id: id }).update({
```
Comment (bodinsamuel) [inline packages/shared/lib/services/external-webhook.service.ts]: At no point we seems create a row for `_nango_external_webhooks`, it will break for new user or I missed something

File: packages/webhooks/lib/utils.ts [hunk 10]
```diff
@@ -115,17 +146,23 @@ export const deliver = async ({
...
+                    }
                 } else {
-                    await logCtx?.error(
+                    await logCtx.error(
                         `${webhookType} sent webhook successfully to the ${type} ${url} but received a ${response.status} response code${endingMessage ? ` ${endingMessage}` : ''}. Please send a 2xx on successful receipt.`,
                         body as Record<string, unknown>
                     );
+                    if (webhookType === 'forward') {
+                        await logCtx.failed();
+                    }
```
Comment (bodinsamuel) [inline packages/webhooks/lib/utils.ts]: I would advocate for never calling failed/success deep inside methods, it's hard to follow the context everywhere. This function could return a boolean or a res, so you could at least do it closer to the creation

File: packages/webhooks/lib/forward.ts [hunk 11]
```diff
@@ -0,0 +1,83 @@
...
+                ...body,
+                connectionId
+            },
+            webhookType: 'forward',
+            activityLogId: null,
+            environment,
+            logCtx,
+            incomingHeaders: webhookOriginalHeaders
+        });
+    }
```
Comment (bodinsamuel) [inline packages/webhooks/lib/forward.ts:90]: logCtx is never success/failed

**Review by bodinsamuel — COMMENTED** [review]
Works locally 📤
Minor comments

File: packages/node-client/lib/types.ts (hunk 2 above)
Comment (khaliqgant) [reply packages/node-client/lib/types.ts:7]: Updated

File: packages/shared/lib/services/external-webhook.service.ts (hunk 3 above)
Comment (khaliqgant) [reply packages/shared/lib/services/external-webhook.service.ts]: Updated.

File: packages/server/lib/controllers/v1/environment/webhook/settings.ts (hunk 4 above)
Comment (khaliqgant) [reply packages/server/lib/controllers/v1/environment/webhook/settings.ts]: Uodated

File: packages/server/lib/routes.ts (hunk 5 above)
Comment (khaliqgant) [reply packages/server/lib/routes.ts]: Fair, but kept it the same.

File: packages/types/lib/environment/api/webhook.ts (hunk 6 above)
Comment (khaliqgant) [reply packages/types/lib/environment/api/webhook.ts:33]: Yeah, good point. Updated

File: packages/server/lib/controllers/v1/environment/webhook/url.ts (hunk 8 above)
Comment (khaliqgant) [reply packages/server/lib/controllers/v1/environment/webhook/url.ts]: Thanks, updated

File: packages/shared/lib/services/external-webhook.service.ts (hunk 9 above)
Comment (khaliqgant) [reply packages/shared/lib/services/external-webhook.service.ts]: Great catch! Updated.

File: packages/webhooks/lib/utils.ts (hunk 10 above)
Comment (khaliqgant) [reply packages/webhooks/lib/utils.ts]: Updated

File: packages/webhooks/lib/forward.ts (hunk 11 above)
Comment (khaliqgant) [reply packages/webhooks/lib/forward.ts:90]: Updated with the above https://github.com/NangoHQ/nango/pull/2319#discussion_r1639592905

**Review by bodinsamuel — APPROVED** [review]

File: packages/node-client/package.json [hunk 12]
```diff
@@ -36,6 +36,7 @@
         "README.md"
     ],
     "devDependencies": {
+        "@nangohq/types": "^0.40.0",
```
Comment (TBonnin) [inline packages/node-client/package.json:39]: should we use `../types` for devDependencies?

File: packages/webhooks/lib/sync.ts [hunk 13]
```diff
@@ -1,101 +1,114 @@
...
-        body.startedAt = dayjs(now).toDate().toISOString();
-        body.failedAt = new Date().toISOString();
-    }
+        successBody = {
+            ...body,
+            success: true,
+            responseResults: {
+                added: responseResults.added,
+                updated: responseResults.updated,
+                deleted: 0
```
Comment (TBonnin) [inline packages/webhooks/lib/sync.ts:84]: why `deleted: 0` and not `deleted: responseResults.deleted || 0`

**Review by TBonnin — APPROVED** [review]

File: packages/node-client/package.json (hunk 12 above)
Comment (khaliqgant) [reply packages/node-client/package.json:39]: Since node-client is published we have to use the published version and not the local one

Comment (TBonnin) [reply packages/node-client/package.json:39]: does `npm install` also install the devDependencies?

Comment (khaliqgant) [reply packages/node-client/package.json:39]: Yes, unless you pass a `--production` flag or if `NODE_ENV` environment variable is set to `production`

File: packages/webhooks/lib/sync.ts (hunk 13 above)
Comment (khaliqgant) [reply packages/webhooks/lib/sync.ts:84]: fair, but i'll leave as is
